flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java

```diff
@@ -313,9 +313,19 @@ protected void resubmitJob(FlinkResourceContext<CR> ctx, boolean requireHaMetada
             throws Exception {
         LOG.info("Resubmitting Flink job...");
         SPEC specToRecover = ReconciliationUtils.getDeployedSpec(ctx.getResource());
+        Optional<Savepoint> lastSavepoint =
+                Optional.ofNullable(
+                        ctx.getResource()
+                                .getStatus()
+                                .getJobStatus()
+                                .getSavepointInfo()
+                                .getLastSavepoint());
         if (requireHaMetadata) {
             specToRecover.getJob().setUpgradeMode(UpgradeMode.LAST_STATE);
+        } else if (lastSavepoint.isPresent()) {
```

Comment:
I wonder if we should change this to `ctx.getResource().getSpec().getJob().getUpgradeMode() != STATELESS && lastSavepoint.isPresent()`

Wouldn't it be weird that you have stateless upgrade mode (for a reason) and maybe a savepoint from a very long time ago and the job still gets submitted from it?